File: longest_substring/longestSubstring.py
from typing import *

# Longest Substring Without Repeating Characters

class Solution:
    def longestSubstring(self, s: str) -> int:
        if not s:
            return 0
        l, r = 0, 0
        res = 0
        seen = set()
        while r < len(s):
            if s[r] not in seen:
                seen.add(s[r])
                res = max(res, r - l + 1) # Update the result if the current substring is longer than the previous one
                r += 1
            else:
                seen.remove(s[l])   # Remove the leftmost character from the set if it is already in the set
                l += 1
        return res

# Test the function
solution = Solution()
s = "abcabcbb"
result = solution.longestSubstring(s)
print(result)  # Output should be 3

File: longest_substring/longest_substring.go
package main

import "fmt"

func longest_substring(s string) int {
	if len(s) == 0 {
		return 0
	}
	l, r := 0, 0
	res := 0
	seen := map[byte]int{}

	for r < len(s) {
		if _, ok := seen[s[r]]; ok {
			l = max(l, seen[s[r]]+1)
		} else {
			res = max(res, r-l+1)
		}
		seen[s[r]] = r
		r++
	}
	return res
}

func main() {
	s := "abcabcbb"
	fmt.Println(longest_substring(s))
}
